### ISSUE TITLE
Return value from find-transactions callback

### ISSUE THREAD

**Issue by babouinette** [transducer/cljs-iota#1]
First let me say I really like using cljs-iota in cljs and I had no problem using it until this.

I'm having trouble with returning a value from a callback. 

My code is:

(defn find-transactions 
    "search transactions associated with an address and return a vector of hashes" 
    []
            (println "finding transactions")
            (iota-api/find-transactions iota-instance {:addresses (list address-to-search)} find-transactions-callback)
          )

And the callback function is:

(defn find-transactions-callback [err hashes]
          (prn  "transactions found" hashes)
          hashes
    )

I expect this to return a vector of hashes, but i got undefined.

If I do
(.log js/console "hashes are" (find-transactions))

The callback prints to the console :

"transactions found" ["JW9QTNHDLGQYKGHLYHWHYRKTTEVUVBZSHNMLCDLBWVYPROKTCDRAMFGXMIQCEM9YIRXUSZJTEKG999999" "QYKZWELJYAMWPITSTZSXLFFERSGRPBKIOCHSE9KLENRXNBXSGSLBEYW9JUVJYP9QWBJJFKMSUDA999999"]
which means it retrieves the result that I want, but console.log prints

"hashes are" undefined

I don' t know why it doesn't return the vector of hashes.

PS: I can't return anything from the callback. Even doing

(defn find-transactions-callback [err hashes]
          "hello"
    )

doesnt work. If I do ((.log js/console "hashes are" (find-transactions)) i got " hashes are" nil.

PS2 : I do exactly the same with get-new-address and in the callback i do 
(fn [err address] (first address))
and it returns the first address of the vector.

Do you have any idea why?

**Comment by transducer** [comment]
>First let me say I really like using cljs-iota in cljs and I had no problem using it until this.

Good to hear! Thanks!

In `get-new-address` the underlying `iota.lib.js`'s `getNewAddress` is called which [returns the callback](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/api/api.js#L839) (so then if the callback returns it will return that returned value). 

In `find-transactions` the underlying `iota.lib.js`'s `findTransactions` is called [which dispatches the callback to `sendCommand`](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/api/api.js#L178), [which dispatches to `_makeRequest`'s  `send`](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/api/api.js#L57), and `_makeRequest` is [equal to the provider`s makeRequest](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/api/api.js#L19), [which is initialized in iota.js and [comes from `makeRequest.js`](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/iota.js#L36),  and the [`send` in `_makeRequest` dispatches to `prepareResult`](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/utils/makeRequest.js#L104), and `makeRequest`'s [`prepareResult` DOES return the callback](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/utils/makeRequest.js#L351).

So in the case of `find-transactions` you should also be able to return a value via a callback in this case since the callback is returned. Maybe somewhere it goes wrong... I don't know.

Anyhow, if you want do want to get the value out of the callback you can for example use a core.async channel:

```

;; Import core.async stuff
(ns something
  (:require-macros [cljs.core.async.macros :refer [go]])
  (:require [cljs-iota.api :as iota-api]
            [cljs-iota.core :as iota]
            [cljs.core.async :as async :refer [<! >! chan]]))


(def iota-instance (iota/create-iota "https://iota-nod3.de/"))
(def address-to-search "CLVZLOKLTDQMHVFPOE9YJHO9IUQMENGZWPXXDMBTZLGXRS9D9AQFEABHIJQJYDHDLWLDESXVM9UJTGIYZTEWICUOWX")


(defn find-transactions []
  (go
    (let [c (chan)]
      (iota-api/find-transactions iota-instance
                                  {:addresses (list address-to-search)}
                                  (fn [err hashes]
                                    (prn "transactions found" hashes)
                                    (go (>! c hashes))))
      (prn "hashes are" (<! c)))))


(find-transactions)
```

Which prints both the transactions and hashes.

In the above code first a core.async channel is created. Then in the callback a value is put (`>!`) on that core.async channel. And finally it is taken (`<!`) off the channel and printed.

Hope this helps.

**Comment by transducer** [comment]
>First let me say I really like using cljs-iota in cljs and I had no problem using it until this.

Good to hear! Thanks!

In `get-new-address` the underlying `iota.lib.js`'s `getNewAddress` is called which [returns the callback](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/api/api.js#L839) (so then if the callback returns it will return that returned value). 

In `find-transactions` the underlying `iota.lib.js`'s `findTransactions` is called [which dispatches the callback to `sendCommand`](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/api/api.js#L178), [which dispatches to `_makeRequest`'s  `send`](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/api/api.js#L57), and `_makeRequest` is [equal to the provider's makeRequest](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/api/api.js#L19), [which is initialized in iota.js and comes from `makeRequest.js`](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/iota.js#L36),  and the [`send` in `_makeRequest` dispatches to `prepareResult`](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/utils/makeRequest.js#L104), and `makeRequest`'s [`prepareResult` does return the callback](https://github.com/iotaledger/iota.lib.js/blob/3e451efc079333c28a9f6cc3d38e13d44667a787/lib/utils/makeRequest.js#L351).

So in the case of `find-transactions` you should also be able to return a value via a callback in this case since the callback is returned. Maybe somewhere it goes wrong... I don't see it.

Anyhow, if you want do want to get the value out of the callback you can for example use a core.async channel:

```

;; Import core.async stuff
(ns something
  (:require-macros [cljs.core.async.macros :refer [go]])
  (:require [cljs-iota.api :as iota-api]
            [cljs-iota.core :as iota]
            [cljs.core.async :as async :refer [<! >! chan]]))


(def iota-instance (iota/create-iota "https://iota-nod3.de/"))
(def address-to-search "CLVZLOKLTDQMHVFPOE9YJHO9IUQMENGZWPXXDMBTZLGXRS9D9AQFEABHIJQJYDHDLWLDESXVM9UJTGIYZTEWICUOWX")


(defn find-transactions []
  (go
    (let [c (chan)]
      (iota-api/find-transactions iota-instance
                                  {:addresses (list address-to-search)}
                                  (fn [err hashes]
                                    (prn "transactions found" hashes)
                                    (go (>! c hashes))))
      (prn "hashes are" (<! c)))))


(find-transactions)
```

Which prints both the transactions and hashes.

In the above code first a core.async channel is created. Then in the callback a value is put (`>!`) on that core.async channel. And finally it is taken (`<!`) off the channel and printed.

Hope this helps.

**Comment by babouinette** [comment]
Hi. 

Thank you so much for your detailed answer. It's so informative. 

I tried to look at the code and didn't understand why it would't return, that's why I was blocked...

I got it working by using a re-frame/dispatch in the callback and storing the hashes in the app-state, and then subscibing to it. It works really well.

Next time if I need to I will try to use async.

Thank you for looking at it :)
